api-get.go

```diff
@@ -207,78 +188,141 @@ func (c Client) GetObjectPartial(bucketName, objectName string) (ReadAtCloser, O
 }
 
-// objectReadAtCloser container for io.ReadAtCloser.
-type objectReadAtCloser struct {
+// Object represents an open object.
+type Object struct {
 	// mutex.
 	mutex *sync.Mutex
 
 	// User allocated and defined.
-	reqCh      chan<- readAtRequest
-	resCh      <-chan readAtResponse
-	doneCh     chan<- struct{}
-	objectSize int64
+	reqCh         chan<- readRequest
+	resCh         <-chan readResponse
+	doneCh        chan<- struct{}
+	currentOffset int64
+	objectInfo    ObjectInfo
 
 	// Previous error saved for future calls.
 	prevErr error
 }
 
-// newObjectReadAtCloser implements a io.ReadSeeker for a HTTP stream.
-func newObjectReadAtCloser(reqCh chan<- readAtRequest, resCh <-chan readAtResponse, doneCh chan<- struct{}, objectSize int64) *objectReadAtCloser {
-	return &objectReadAtCloser{
-		mutex:      new(sync.Mutex),
-		reqCh:      reqCh,
-		resCh:      resCh,
-		doneCh:     doneCh,
-		objectSize: objectSize,
+// Read reads up to len(p) bytes into p. It returns the number of bytes read
+// (0 <= n <= len(p)) and any error encountered. returns io.EOF upon end of file.
+func (o *Object) Read(b []byte) (n int, err error) {
+	if o == nil {
+		return 0, ErrInvalidArgument("Object is nil")
+	}
+
+	// Locking.
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if o.currentOffset >= o.objectInfo.Size {
+		return 0, io.EOF
+	}
+
+	// prevErr is which was saved in previous operation.
+	if o.prevErr != nil {
+		return 0, o.prevErr
 	}
+
+	// Send current information over control channel to indicate we are ready.
+	reqMsg := readRequest{}
+
+	// Send the current offset and bytes requested.
+	reqMsg.Buffer = b
+	reqMsg.Offset = o.currentOffset
+
+	// Send read request over the control channel.
+	o.reqCh <- reqMsg
+
+	// Get data over the response channel.
+	dataMsg := <-o.resCh
+
+	// Bytes read.
+	bytesRead := int64(dataMsg.Size)
+
+	// update current offset.
+	o.currentOffset += bytesRead
+
+	if dataMsg.Error == nil {
+		// If currentOffset read is equal to objectSize
+		// we have reached end of file, we return io.EOF.
+		if o.currentOffset >= o.objectInfo.Size {
+			return dataMsg.Size, io.EOF
+		}
+		return dataMsg.Size, nil
+	}
+
+	// Save any error.
+	o.prevErr = dataMsg.Error
+	return dataMsg.Size, dataMsg.Error
+}
+
+// Stat returns the ObjectInfo structure describing object.
+func (o *Object) Stat() (ObjectInfo, error) {
+	if o == nil {
+		return ObjectInfo{}, ErrInvalidArgument("Object is nil")
+	}
+	// Locking.
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if o.prevErr != nil {
+		return ObjectInfo{}, o.prevErr
+	}
+
+	return o.objectInfo, nil
 }
 
 // ReadAt reads len(b) bytes from the File starting at byte offset off.
 // It returns the number of bytes read and the error, if any.
 // ReadAt always returns a non-nil error when n < len(b).
 // At end of file, that error is io.EOF.
-func (r *objectReadAtCloser) ReadAt(b []byte, offset int64) (int, error) {
+func (o *Object) ReadAt(b []byte, offset int64) (n int, err error) {
```

Comment:
I thought we are going to implement Seek over ReadAt. I will change progressbar code accordingly.